### PR TITLE
some block device handling fixes

### PR DESCRIPTION
**Pull request by poettering**
I noticed these while reviewing #9247. I figure that PR should be rebased on this as soon as it is merged.

### CODE REVIEW

File: src/basic/blockdev-util.c [hunk 1]
```diff
@@ -85,35 +85,22 @@ int get_block_device(const char *path, dev_t *dev) {
...
-
-        /* Gets the backing block device for a file system, and
-         * handles LUKS encrypted file systems, looking for its
-         * immediate parent, if there is one. */
-
-        r = get_block_device(path, &dt);
-        if (r <= 0)
-                return r;
+        /* For the specified block device tries to chase it down the layers, in case LUKS-style DM stacking is used,
+         * trying to find the next upper layer.  */
```
Comment (keszybz) [inline src/basic/blockdev-util.c]: down the layers to go up?

File: src/core/cgroup.c [hunk 2]
```diff
@@ -306,30 +307,36 @@ void cgroup_context_dump(CGroupContext *c, FILE* f, const char *prefix) {
...
+                *ret = st.st_dev; /* If this is not a device node then use the block device this file is stored on */
+        else {
+                /* If this is btrfs, getting the backing block device is a bit harder */
+                r = btrfs_get_block_device(p, ret);
+                if (r == -ENOTTY) {
+                        log_warning("'%s' is not a block device node, and file system block device cannot be determined or is not local.", p);
+                        return -ENODEV;
+                }
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to determine block device backing btrfs file system '%s': %m", p);
```
Comment (keszybz) [inline src/core/cgroup.c]: This message assumes that `p` must be a btrfs device. I think it'd be better to reorder this and report the more specific case of failing to determine btrfs backing device first:
```C
                r = btrfs_get_block_device(p, ret);
                if (r < 0 && r != -ENOTTY)
                        return log_warning_errno(r, "Failed to determine if path '%s' is backed by a btrfs file system: %m", p);

                if (r == -ENOTTY) {
                        log_warning("'%s' is not a block device node, and file system block device cannot be determined or is not local.", p);
                        return -ENODEV;
                }
```

File: src/basic/blockdev-util.c (hunk 1 above)
Comment (poettering) [reply src/basic/blockdev-util.c]: up, down, potato, potato ;-)

**Comment by poettering** [comment]
Force pushed a new version now. PTAL. Only changes are the points you raised

**Comment by keszybz** [comment]
LGTM.

**Pull request merged**